### PR TITLE
Changes for Nginx 1.7.7

### PR DESCRIPTION
**Pull request by julek-wolfssl**
Push error when decryption fails
If wolfSSL_CTX_use_certificate keeps passed in cert then it should either copy it or increase its reference counter

### CODE REVIEW

File: src/ssl.c [hunk 1]
```diff
@@ -5522,8 +5522,14 @@ int ProcessBuffer(WOLFSSL_CTX* ctx, const unsigned char* buff,
 
         if (ret != 0)
             return ret;
-        if (keyFormat == 0)
+        if (keyFormat == 0) {
+#ifdef OPENSSL_EXTRA
+            /* Reaching this point probably means that the
+             * decryption password is wrong */
+            EVPerr(0, EVP_R_BAD_DECRYPT);
```
Comment (SparkiDev) [inline src/ssl.c:5529]: Check for password set?

File: src/ssl.c [hunk 2]
```diff
@@ -29435,6 +29440,7 @@ int SetDhExternal(WOLFSSL_DH *dh)
         return WOLFSSL_FATAL_ERROR;
     }
 
+#if !defined(HAVE_FIPS) || (defined(HAVE_FIPS_VERSION) && (HAVE_FIPS_VERSION>2))
```
Comment (SparkiDev) [inline src/ssl.c:29443]: Have you tested this change?

**Review by SparkiDev — CHANGES_REQUESTED** [review]

File: src/ssl.c (hunk 2 above)
Comment (julek-wolfssl) [reply src/ssl.c:29443]: Yes.

**Review by SparkiDev — APPROVED** [review]

File: src/ssl.c [hunk 3]
```diff
@@ -29435,6 +29441,7 @@ int SetDhExternal(WOLFSSL_DH *dh)
...
```
Comment (toddouska) [inline src/ssl.c:29444]: Why is this gated by FIPSv2?  And why only this part of the function?  And why are we returning success to the user in FIPSv1 case??

Comment (julek-wolfssl) [reply src/ssl.c:29444]: The `wolfSSL_PEM_read_bio_DHparams` function was missing from the nginx build using the client's FIPS configuration. When I included it in the FIPS build `SetDhInternal` was missing. When I included `SetDhInternal`, gcc reported that `pub` and `priv` are missing. These members of the `DhKey` struct are not available in the FIPS build but the rest of the struct is available so there is no reason why the entire function should be gated off in the FIPS build.

In short: `pub` and `priv` are not available in FIPS builds but the rest of the `DhKey` struct is available so only the part of function that deals with those members needs to be excluded.

Comment (toddouska) [reply src/ssl.c:29444]: Other areas of the code appear to use  WOLFSSL_DH_EXTRA to gate those variables.  Shouldn't we use the same pattern here instead of FIPS defines?

Comment (julek-wolfssl) [reply src/ssl.c:29444]: It does make more sense to use `WOLFSSL_DH_EXTRA`. I changed it to use that instead.

**Review by toddouska — APPROVED** [review]
